### ISSUE TITLE
OpenSUSE Leap 15.4 chroot initialization hangs

### ISSUE THREAD

**Issue by praiskup** [rpm-software-management/mock#915]
Bootstrap chroot installs just fine, but then `patterns-devel-base-devel_rpm_build` package (buildroot preparation?) fails:
```
Start: dnf install
Failed to set locale, defaulting to C
No matches found for the following disable plugin patterns: local, spacewalk, versionlock
openSUSE Leap 15.4 - x86_64 - OSS                                                                                                                                                                              16 MB/s |  87 MB     00:05    
openSUSE Leap 15.4 - x86_64 - OSS - Updates                                                                                                                                                                    37 kB/s |  14 kB     00:00    
openSUSE Leap 15.4 - x86_64 - Updates from SUSE Linux Enterprise                                                                                                                                               10 MB/s | 7.8 MB     00:00    
openSUSE Leap 15.4 - x86_64 - Updates from Backports for SUSE Linux Enterprise                                                                                                                                119 kB/s |  52 kB     00:00    
Last metadata expiration check: 0:00:01 ago on Fri Jun 10 10:10:53 2022.
allow_vendor_change is disabled. This option is currently not supported for downgrade and distro-sync commands
Dependencies resolved.
==============================================================================================================================================================================================================================================
 Package                                                                 Architecture                        Version                                                              Repository                                             Size
==============================================================================================================================================================================================================================================
Installing:
 patterns-devel-base-devel_rpm_build                                     x86_64                              20170319-lp154.1.2                                                   opensuse-leap-oss                                     8.6 k
Installing dependencies:
 aaa_base                                                                x86_64                              84.87+git20180409.04c9dae-3.57.1                                     opensuse-leap-oss                                     106 k
 bash                                                                    x86_64                              4.4-150400.25.22                                                     opensuse-leap-oss                                     522 k
 bash-sh                                                                 x86_64                              4.4-150400.25.22                                                     opensuse-leap-oss                                      32 k
...
(204/206): patch-2.7.6-150000.5.3.1.x86_64.rpm                                                                                                                                                                1.6 MB/s | 118 kB     00:00    
(205/206): libtsan0-11.3.0+git1637-150000.1.9.1.x86_64.rpm                                                                                                                                                    3.1 MB/s | 329 kB     00:00    
(206/206): pam-1.3.0-150000.6.58.3.x86_64.rpm                                                                                                                                                                 2.8 MB/s | 435 kB     00:00    
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                                                                                                          12 MB/s | 104 MB     00:08     
warning: Generating 18 missing index(es), please wait...
openSUSE Leap 15.4 - x86_64 - OSS                                                                                                                                                                             965 kB/s | 988  B     00:00    
Importing GPG key 0x3DBDC284:
 Userid     : "openSUSE Project Signing Key <opensuse@opensuse.org>"
 Fingerprint: 22C0 7BA5 3417 8CD0 2EFE 22AA B88B 2FD4 3DBD C284
 From       : /usr/share/distribution-gpg-keys/opensuse/RPM-GPG-KEY-openSUSE
Key imported successfully
openSUSE Leap 15.4 - x86_64 - OSS                                                                                                                                                                             1.1 MB/s | 1.1 kB     00:00    
Importing GPG key 0x65176565:
 Userid     : "openSUSE:Backports OBS Project <openSUSE:Backports@build.opensuse.org>"
 Fingerprint: 637B 32FF 3D83 F07A 7AE1 C40A 9C21 4D40 6517 6565
 From       : /usr/share/distribution-gpg-keys/opensuse/RPM-GPG-KEY-openSUSE-Backports
Key imported successfully
openSUSE Leap 15.4 - x86_64 - OSS                                                                                                                                                                             915 kB/s | 937  B     00:00    
Importing GPG key 0x39DB7C82:
 Userid     : "SuSE Package Signing Key <build@suse.de>"
 Fingerprint: FEAB 5025 39D8 46DB 2C09 61CA 70AF 9E81 39DB 7C82
 From       : /usr/share/distribution-gpg-keys/suse/RPM-GPG-KEY-SuSE-SLE-15
Key imported successfully
warning: waiting for shared lock on /var/lib/mock/opensuse-leap-15.4-x86_64/root/usr/lib/sysimage/rpm/Packages
^CWARNING: Machine f3dfa12f49d742179f04494209a5f99a still running. Killing...
ERROR: Exiting on user interrupt, <CTRL>-C
```

**Comment by praiskup** [comment]
@Conan-Kudo any idea?

**Comment by Conan-Kudo** [comment]
Hmm, I can reproduce this on Fedora 36.

Here are the full errors I see:

```
warning: waiting for shared lock on /var/lib/mock/opensuse-leap-15.4-x86_64/root/usr/lib/sysimage/rpm/Packages
error: cannot get shared lock on /var/lib/mock/opensuse-leap-15.4-x86_64/root/usr/lib/sysimage/rpm/Packages
error: cannot open Packages index using db4 - Operation not permitted (1)
error: cannot open Packages database in /var/lib/mock/opensuse-leap-15.4-x86_64/root/usr/lib/sysimage/rpm
Import of key(s) didn't help, wrong key(s)?
warning: waiting for shared lock on /var/lib/mock/opensuse-leap-15.4-x86_64/root/usr/lib/sysimage/rpm/Packages
```

@lkocman, @dirkmueller, @mlschroe have you heard of anything like this?

**Comment by Conan-Kudo** [comment]
What's bizarre is that killing it and running it again fixes it...

**Comment by Conan-Kudo** [comment]
The only thing I can see that might be related would be the following changelog entry in the rpm package:

```rpmspec
* Wed Oct 06 2021 mls@suse.de
  - backport zstd detection fix [bsc#1187670]
    new patch: zstddetection.diff
  - backport ndb rofs support [bsc#1188548]
    new patch: ndbrofs.diff
  - backport pgp hardening changes from upstream [bsc#1185299]
    new patch: pgpharden.diff
  - fix deadlock when multiple rpm processes try tp acquire the
    database lock [bsc#1183659]
    new patch: deadlock.diff
```

Notably, the deadlock changelog entry here, which references a private bug. @lkocman, @mlschroe: could either of you provide more information on [bsc#1183659](https://bugzilla.suse.com/1183659)?

**Comment by praiskup** [comment]
Ping on this, still the same issue.  Should we keep 15.4 configs?

**Comment by brianjmurrell** [comment]
> Ping on this, still the same issue. Should we keep 15.4 configs?

A month later, ping again.  What's the path forward here?  I need to start doing Leap 15.4 builds.  Is there a temporary workaround even?

**Comment by Conan-Kudo** [comment]
If you run it twice, it seems to fix itself.

**Comment by brianjmurrell** [comment]
> If you run it twice, it seems to fix itself.

While that's a reasonable work-around for a user at a keyboard that can `^C` the hung process and run it again, it doesn't fit very well or nicely into a CI system unfortunately.  Is there nothing else that can prevent this the way running it that first time does?

**Comment by Conan-Kudo** [comment]
It's a SUSE problem. I don't know if anyone has filed a ticket with them about it.

**Comment by brianjmurrell** [comment]
To the point of filing at ticket with SUSE, Is there a simpler reproducer than running `mock`?

**Comment by brianjmurrell** [comment]
Ultimately what seems to be happening is that `dnf` and it's child `rpmkeys` are deadlocking:

```
$ sudo fuser /var/lib/mock/opensuse-leap-15.4-x86_64/root/usr/lib/sysimage/rpm/Packages
/var/lib/mock/opensuse-leap-15.4-x86_64/root/usr/lib/sysimage/rpm/Packages: 379324 379680
$ ps -p379324,379680 fwwwww
    PID TTY      STAT   TIME COMMAND
 379324 ?        Ss     0:02 /usr/bin/python3 /usr/bin/dnf --installroot /var/lib/mock/opensuse-leap-15.4-x86_64/root/ --releasever 15.4 --disablerepo * --enablerepo *-artifactory --disablerepo *-debug* --disablerepo sclo* --setopt=deltarpm=False --allowerasing --disableplugin=local --disableplugin=spacewalk --disableplugin=versionlock install patterns-devel-base-devel_rpm_build --setopt=tsflags=nocontexts
 379680 ?        S      0:00  \_ rpmkeys --checksig --root /var/lib/mock/opensuse-leap-15.4-x86_64/root/ --verbose --define=_pkgverify_level signature --define=_pkgverify_flags 0x0 -
```

Are we sure this is something SUSE specific?

**Comment by brianjmurrell** [comment]
> If you run it twice, it seems to fix itself.

This is not always my experience.  I have been able to hit this failure many times in a row simply trying to rerun the same `mock` command in my attempts to automate this suggested work-around.

**Comment by Conan-Kudo** [comment]
Here's an openSUSE bug filed by @lkocman: https://bugzilla.opensuse.org/show_bug.cgi?id=1202750

**Comment by xsuchy** [comment]
I will leave this issue open, but I do not expect anyone from Mock team working on this as I cannot imagine Mock being a culprit. I will leave it open so people do not open duplicate. But this needs to be fixed on openSUSE side.

**Comment by mlschroe** [comment]
I think this is caused by dnf using a call to rpmkeys for signature verification (i.e. commit https://github.com/rpm-software-management/dnf/commit/4747e369c1b5b406688ff0be5447ebd0c29575a6). SUSE uses a different database locking scheme, and importing a key via self._ts.pgpImportPubkey() will switch the database to "exclusive locked". After that forking and calling rpmkeys will result in a deadlock.

**Comment by brianjmurrell** [comment]
So what's the solution here?

**Comment by mlschroe** [comment]
I'll release a maintenance update for rpm that makes it go back the a read-only lock after importing a key if it started read-only. I guess that's the easiest solution.

**Comment by dmach** [comment]
@pmatilai any idea if rpm provides sufficient python API so we could use it instead of calling out rpmkeys?
I'm interested in the latest rpm version and rpm-4.14.3 that is in openSUSE Leap 15.4.

@j-mracek @jrohel just out of curiosity - how dnf5 handles this? (maybe using C API solves this)

**Comment by brianjmurrell** [comment]
> I'll release a maintenance update for rpm that makes it go back the a read-only lock after importing a key if it started read-only. I guess that's the easiest solution.

You are referring to a maintenance update for SLES/Leap 15.4, correct?  Any progress on this?  I'd be happy to test a pre-release build if it's available.

**Comment by brianjmurrell** [comment]
This appears to have been resolved in a recent update to Leap 15.4.

**Comment by praiskup** [comment]
Indeed, this is resolved.  Leap 15.4 and 15.5 seem to install and build fine with Mock 5.5 on Fedora.
